modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java

```diff
@@ -60,13 +60,29 @@
  * results.
  */
 public abstract class AbstractAsyncBulkIndexByScrollAction<Request extends AbstractBulkByScrollRequest<Request>>
-        extends AbstractAsyncBulkByScrollAction<Request>
-        implements BiFunction<ActionRequest<?>, SearchHit, ActionRequest<?>> {
+        extends AbstractAsyncBulkByScrollAction<Request> {
+
+    protected final ScriptService scriptService;
+    protected final ClusterState clusterState;
+    private final ScriptApplier scriptApplier;
 
     public AbstractAsyncBulkIndexByScrollAction(BulkByScrollTask task, ESLogger logger, ParentTaskAssigningClient client,
                                                 ThreadPool threadPool, Request mainRequest, SearchRequest firstSearchRequest,
-                                                ActionListener<BulkIndexByScrollResponse> listener) {
+                                                ActionListener<BulkIndexByScrollResponse> listener,
+                                                ScriptService scriptService, ClusterState clusterState) {
         super(task, logger, client, threadPool, mainRequest, firstSearchRequest, listener);
+        this.scriptService = scriptService;
+        this.clusterState = clusterState;
+        this.scriptApplier = getScriptApplier();
+    }
+
+    /**
+     * Return the {@link ScriptApplier} to apply to all {@link ActionRequest}, null if
+     * no {@link ScriptApplier} must be applied.
+     */
+    protected ScriptApplier getScriptApplier() {
```

Comment:
Rename to `buildScriptApplier` and return a noop implementation by default? I usually prefer that to null checks for things that behave like this.